src/checkers/inference/dataflow/InferenceTransfer.java

```diff
@@ -401,4 +408,94 @@ private InferenceAnalysis getInferenceAnalysis() {
     private boolean isDeclarationWithInitializer(AssignmentNode assignmentNode) {
         return (assignmentNode.getTree().getKind() == Tree.Kind.VARIABLE);
     }
+
+    private void createComparisonVariableSlot(Node node, CFStore thenStore, CFStore elseStore) {
+        // Only create refinement comparison slot for variables
+        // TODO: deal with comparison between more complex expressions
+        Node var = node;
+        if (node instanceof AssignmentNode) {
+            AssignmentNode a = (AssignmentNode) node;
+            var = a.getTarget();
+        }
+        if (!(var instanceof LocalVariableNode) && !(var instanceof FieldAccessNode)) {
+            return;
+        }
+        Tree tree = var.getTree();
+        ConstraintManager constraintManager = InferenceMain.getInstance().getConstraintManager();
```

Comment:
I think it works in either ways. I guess putting constraint creation here is just for convenience. I can move it to `InferenceVisitor.visitBinary` for consistency. In that case, we should (1) check if it's a comparison (2) retrieve the comparison slots associated from the cache.

Reply:
Not all the constraints are created in `xxxVisitor`. E.g., the equality constraint associated an assignment is created in the dataflow transfer function, as https://github.com/opprop/checker-framework-inference/blob/21dfba4fd0013dd97f0c477e828cd9dbda4b6356/src/checkers/inference/DefaultSlotManager.java#L519

For refinement variables we add the subtype constraints in InferenceVisitor may be because we want to reuse the super method `BaseTypeVisitor.visitAssignment`, which checks if RHS <: LHS.

Besides, regarding the comparison variable slot, I think the subtype constraints (compareSlot <: declareType) are not precise. It's more precise to create compareSlot <: typeBeforeComparison, i.e. line 443-446 can be removed.

Reply:
Besides, regarding the comparison variable slot, I think the subtype constraints (compareSlot <: declareType) are not precise. It's more precise to create compareSlot <: typeBeforeComparison, i.e. line 443-446 can be removed.